### DISCUSSION TITLE
4.1.0.beta New features and fixes

### DISCUSSION THREAD

**Discussion by seb5g** [PyMoDAQ/PyMoDAQ#195] [General]
## What's Changed
* make sure the averaging within daq_viewer is applied when using daq_scan by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/152
* Scan close by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/155
* close properly the parent widget (mainwindow) by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/156
* make sure one can now use SpecialSlicerData to set part of a dataobject by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/158
* replacing the dwa "origin" by the control module's title by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/160
* make sure the total dim included nav axes does not exceed 2D by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/161
* Add plugin package name to documentation by @bmoneke in https://github.com/PyMoDAQ/PyMoDAQ/pull/162
* Fix node exporters by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/163
* redefines the Vertical label class by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/165
* reverting to using pyqtgraph stuff (using the actual PR#2850 in pyqtg… by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/166
* make sure one can resize the header of the parametertree by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/167
* publishing dedicated doc on the use/methods to deal with axis in an actuator by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/171
* Update instrument_plugins.rst by @bmoneke in https://github.com/PyMoDAQ/PyMoDAQ/pull/164
* Fix 168 threading problems by @bmoneke in https://github.com/PyMoDAQ/PyMoDAQ/pull/174
* Documentation for Linux installation  by @Ashwolaa in https://github.com/PyMoDAQ/PyMoDAQ/pull/182
* TCP/IP for PyMoDAQ4 - Serialization/Deserialization by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/183
* updated the tcp/ip communication documentation by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/184
* Deepcopy of dte to prevent change of axis indices by daq_scan by @rgeneaux in https://github.com/PyMoDAQ/PyMoDAQ/pull/178
* adds overlay into 1D dataviewer by @seb5g in https://github.com/PyMoDAQ/PyMoDAQ/pull/186
* Update installation.rst by @Ashwolaa in https://github.com/PyMoDAQ/PyMoDAQ/pull/191

## New Contributors
* @bmoneke made their first contribution in https://github.com/PyMoDAQ/PyMoDAQ/pull/162
* @Ashwolaa made their first contribution in https://github.com/PyMoDAQ/PyMoDAQ/pull/182

**Full Changelog**: https://github.com/PyMoDAQ/PyMoDAQ/compare/4.0.8...4.1.0.beta

<hr /><em>This discussion was created from the release <a href='https://github.com/PyMoDAQ/PyMoDAQ/releases/tag/4.1.0.beta'>4.1.0.beta New features and fixes</a>.</em>

**Comment by seb5g** [comment]
I prepared a new branch called 4.1.0 (for the futur release) to be tested (together with beta release 4.1.0.beta). It means that PR related to bug fixes of its new features will be pushed on this branch, while we keep pymodaq_dev for more new functionnalities. You have then 2 possibilities: either install from the 4.1.0 branch or from the beta release: `pip install pymodaq==4.1.0.beta` . If there are things to discuss about it there is a discussion on the subject.

**Comment by quantumm** [comment]
Hi,
I just installed the beta release on a fresh environment with python 3.10 and pyqt5 on Windows 10. When I try to launch a daq_move i get this error:

(pmd4beta) C:\Users\attose10\tmp\pymodaq_plugins_lecroy>daq_move
Traceback (most recent call last):
  File "C:\Users\attose10\Anaconda3\envs\pmd4beta\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\attose10\Anaconda3\envs\pmd4beta\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\attose10\Anaconda3\envs\pmd4beta\Scripts\daq_move.exe\__main__.py", line 4, in <module>
  File "C:\Users\attose10\Anaconda3\envs\pmd4beta\lib\site-packages\pymodaq\control_modules\daq_move.py", line 23, in <module>
    from pymodaq.control_modules.utils import ControlModule
  File "C:\Users\attose10\Anaconda3\envs\pmd4beta\lib\site-packages\pymodaq\control_modules\utils.py", line 64, in <module>
    raise DetectorError('No installed Detector')
pymodaq.utils.exceptions.DetectorError: No installed Detector

**Reply by BenediktBurger** [reply]
I think previously the preset could not load, unless there is a detector, maybe there was this error check added, which is also executed in this `run_module_as_main`.

**Reply by seb5g** [reply]
This is due to a modification of the pymodaq_plugins_mock package I recently modified. The pre-release version of pymodaq cannot see its instruments anymore

It is due to the new features layout applied to plugins:
![image](https://github.com/PyMoDAQ/PyMoDAQ/assets/6609955/139dd9e1-5a9c-46f7-9d9f-d909d8411342)

These allows to precisely define which feature a plugin is exposing. Under the hood it creates specific entry points that pymodaq can look at. In previous plugin version, the entry point was: `pymodaq.plugins` which is now replaced by `pymodaq.instruments`. The pymodaq_dev branch is taking this into account but not the pre-release version.

To solve this, I just published a patch to pymodaq_plugins_mock to still publish an entry point in the form `pymodaq.plugins` but actually one has to modifiy the method used by pymodaq to fetch entry points in the pre-release code (or its branch)

**Reply by rgeneaux** [reply]
I see that for instance v1.1.0 of https://github.com/PyMoDAQ/pymodaq_plugins_physik_instrumente has the same issue. 
Should we add the line 
`else:
entrypoints['pymodaq.plugins'] = f'{SHORT_PLUGIN_NAME} = {PLUGIN_NAME}'  # for backcompatibility`
to all plugins to ensure backcompatibility?

**Comment by seb5g** [comment]
Actually, I'd like to hear you opinion. Due to the intense development activity following the pymodaq days and the ANF, I think we should postpone the release for a time when there will be less development activity, probably in a few month. what do you think?

**Reply by BenediktBurger** [reply]
As there are no test options, it is good to have the beta release for some time.
Maybe some bug fixes (like the threading issue) could be back ported to a 4.0.9 patch release

**Reply by seb5g** [reply]
yes but how can we disentangle the patch contributions from the feature contribution as they have already been merged into pymodaq_dev...
What do you mean there are no test?

**Comment by rgeneaux** [comment]
I think that there are a lot of changes in the 4.1.0 that need to be tested by several of us! The actual release will have to wait before we decide it feels stable enough - hard to say if this will be in a few month or before that. But I'd say that we should focus on bug fixing rather than developing new features, definitely.

**Reply by seb5g** [reply]
For this I think we should wait and include the things we are working on at the moment also. I'm working in the lab with pymodaq_dev branch so should be soon aware of issues... I'll let you know when development slows down, we'll do another pre-release at that time

**Comment by seb5g** [comment]
We could also apply a patch to the current pymodaq latest release to deal
with that. We would only have to do it once.

**Reply by rgeneaux** [reply]
I went with 
`    discovered_plugins = get_entrypoints(group='pymodaq.instruments') + get_entrypoints(group='pymodaq.plugins')`
(line 607 of daq_utils)
which seems to do the job, except that Mock now appears twice with your patch!

**Reply by seb5g** [reply]
that's the idea, but I already patched it in the pymodaq_dev branch, I will just use that. Your solution is actually buggy with python 3.8 because they changed the entry point API in between ... I will do the patch, publish a release, remove the patch in the mock and we should be good for this...

**Comment by seb5g** [comment]
alright it is done:

- pymodaq 4.0.9 is released (with correct entry point selection)
- pymodaq_plugins_mock is back without the patch

**Comment by seb5g** [comment]
Alright....
The plugin_manager actually needs the info about all the plugins that are installed. Like this it will move the installed plugins into the uninstallable list... with the new fetaure it's not working anymore...

I therefore ask of you all to keep in the setup.py a generic entry point: 'pymodaq.plugins' to advertise python about the fact there is a plugin (we don't know its features but for this we use the other entry points):

so make sure to keep this line in your setup.py:

```
entrypoints['pymodaq.plugins'] = f'{SHORT_PLUGIN_NAME} = {PLUGIN_NAME}'  # generic plugin, usefull for the plugin manager
```

Thansk and sorry for the back and forth micmac...
